### PR TITLE
Feature: Server side log in

### PR DESCRIPTION
**Pull request by OleksiiDoroshenko**
Add server-side log in support.

### CODE REVIEW

File: server/src/main/java/io/javaclasses/filehub/web/routes/LogInRoute.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+        } catch (UserNotRegisteredException e) {
+
+            if (logger.isErrorEnabled()) {
+                logger.error(format("Error %s occurred. With message: %s.", e.getClass(), e.getMessage()));
+            }
+
+            response.status(SC_CONFLICT);
+            return e.getMessage();
+        } catch (InvalidUserCredentialsException | JsonParseException e) {
```
[the remark this answers is not in the document]
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/web/routes/LogInRoute.java:81]: If request body is invalid in every way response status will be 400.

File: server/src/main/java/io/javaclasses/filehub/web/deserializers/LogInUserDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+            LogInUser command = new LogInUser(new LoginName(loginValue), new Password(passwordValue));
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("Deserialization completed. Object: " + command.toString() + ".");
+            }
+            return command;
+        } catch (Exception e) {
+            throw new JsonParseException("Can not parse request body.");
+        }
```
[the remark this answers is not in the document]
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/web/deserializers/LogInUserDeserializer.java:51]: No, it is not. It also can throw `NullPointerException` and `ClassCastException`.

File: server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedIdUserRecord.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import com.google.errorprone.annotations.Immutable;
+import io.javaclasses.filehub.storage.Record;
+import io.javaclasses.filehub.storage.userStorage.UserId;
+
+import java.time.LocalDate;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Data structure for saving token in the {@link LoggedInUsersStorage}.
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedIdUserRecord.java:12]:
```suggestion
 * A data structure for saving token. A unit of data of {@link LoggedInUsersStorage}.
```
Try splitting this sentence into two: "What is it?" and "What it does?" and thus you will improve readability.
- [ ] Ok.

File: server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedInUsersStorage.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public Optional<LoggedIdUserRecord> get(Token token) {
+        Optional<LoggedIdUserRecord> record = all().stream().filter(item ->
+                item.id().value().equals(token.value())).findAny();
+
+        if (!record.isPresent()) {
+            return Optional.empty();
+        }
+        if (isTokenExpired(record.get())) {
+            super.remove(record.get().id());
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedInUsersStorage.java:31]: Do we actually need `super` reference? If not, just remove it.

- [ ] Ok.

File: server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedInUsersStorage.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Storage for saving {@link LoggedIdUserRecord}.
+ */
+public class LoggedInUsersStorage extends InMemoryRecordStorage<LoggedIdUserRecord, Token> {
+
+    /**
+     * Returns token record by its value.
+     *
+     * <p>Not return token if its expiration date passed.</p>
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedInUsersStorage.java:17]: But what then it returns? Breaks or returns something else?

- [ ] Ok.

PS. `@return` annotation is the best place to describe what it returns and in which case. Here it is much better to describe what it does.
For example: `Retrieves token record by its value.`

File: server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedInUsersStorage.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+        return record;
+    }
+
+    /**
+     * Checks if token of the record is expired.
+     *
+     * @param record record to be checked.
+     * @return true if it is expired / folder id it is not..
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/storage/tokenStorage/LoggedInUsersStorage.java:42]:
```suggestion
     * @return true if it is expired, false it is not.
```

- [ ] Please, fix.

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java [hunk 7]
```diff
@@ -3,15 +3,16 @@
...
+
 
 /**
  * Storage for saving {@link UserRecord}.
  */
 public class UserStorage extends InMemoryRecordStorage<UserRecord, UserId> {
 
     /**
-     * Returns {@link Optional<UserRecord>} by {@link LoginName}.
+     * Returns {@link Optional <UserRecord>} by {@link LoginName}.
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:15]:
```suggestion
     * Returns {@link Optional<UserRecord>} by {@link LoginName}.
```

- [ ] Ok.

Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:15]: The same comment about `@return` section as above. Please, write in the beginning of JavaDoc what method does.

- [ ] Ok.

File: server/src/main/java/io/javaclasses/filehub/web/ServerTimeZone.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package io.javaclasses.filehub.web;
+
+import jdk.nashorn.internal.ir.annotations.Immutable;
+
+import java.time.Clock;
+import java.time.ZoneId;
+
+/**
+ * Provide specific time zone for FileHub application.
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/web/ServerTimeZone.java:9]: Please make sure that JavaDoc begins with a statement about "What is it?". Here is only answer on the question "What it does?".

- [ ] Ok.

File: server/src/main/java/io/javaclasses/filehub/web/ServerTimeZone.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ */
+@Immutable
+public final class ServerTimeZone {
+    /**
+     * Returns specific time zone.
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/web/ServerTimeZone.java:14]: The same comment about `@return` section as above.

- [ ] Ok.

File: server/src/main/java/io/javaclasses/filehub/web/ServerTimeZone.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.time.Clock;
+import java.time.ZoneId;
+
+/**
+ * Provide specific time zone for FileHub application.
+ */
+@Immutable
+public final class ServerTimeZone {
+    /**
```
Comment (YehorPytomets) [inline server/src/main/java/io/javaclasses/filehub/web/ServerTimeZone.java:13]: Consider adding an empty line between class declaration and its members.

- [ ] Ok.

File: server/src/test/java/io/javaclasses/fileHub/storage/TokenStorage/LoggedInUsersStorageTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import java.time.Period;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+
+@DisplayName("TokenStorage should: ")
+public class LoggedInUsersStorageTest {
+
+    @DisplayName("correctly return existed record by its value.")
```
Comment (YehorPytomets) [inline server/src/test/java/io/javaclasses/fileHub/storage/TokenStorage/LoggedInUsersStorageTest.java:22]:
```suggestion
    @DisplayName("correctly return existing record by its value.")
```
Use `existing` if you mean that the record exists at the moment of test.

- [ ] Ok.

File: server/src/test/java/io/javaclasses/fileHub/storage/TokenStorage/LoggedInUsersStorageTest.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+    @SuppressWarnings("LocalDateTemporalAmount")
+    private Token addValidRecordToStorage(LoggedInUsersStorage storage) {
+        Token id = new Token("test");
+        LoggedIdUserRecord record = new LoggedIdUserRecord(id, new UserId("test"),
+                LocalDate.now(ServerTimeZone.get()).plus(Period.ofDays(3)));
+        return storage.add(record);
+    }
+
+    @DisplayName("not return token if its expiration date was passed.")
```
Comment (YehorPytomets) [inline server/src/test/java/io/javaclasses/fileHub/storage/TokenStorage/LoggedInUsersStorageTest.java:42]:
```suggestion
    @DisplayName("not return token if its expiration date passed.")
```

- [ ] Ok.

**Pull request merged**